src/checks/substitution_key.rs

```diff
@@ -0,0 +1,184 @@
+
+        // Checks if keys used in value have both '{' '}' or neither
+        while let Some(index) = value.find('$') {
+            let prefix = &value[..index];
+            let raw_key = &value[index + 1..];
+
+            // Separate initial key from the rest
+            let (initial_key, rest) = raw_key
+                .find(|c: char| c == '$')
+                .map(|i| (&raw_key[..i], &raw_key[i..]))
+                .unwrap_or_else(|| (raw_key, ""));
+
+            // Ensure if key starts with a '{' that it ends with a '}', else set
+            // bad_substitution to true. Also set key to the proper substitution key.
+            let (key, bad_substitution) = match initial_key.find(|c: char| c == '{') {
+                Some(0) => initial_key
+                    .find(|c: char| c == '}')
+                    .map(|i| (&initial_key[1..i], false))
+                    .unwrap_or_else(|| (&initial_key, true)),
+                _ => initial_key
+                    .find(|c: char| !c.is_ascii_alphanumeric() && c != '_')
+                    .map(|i| (&initial_key[..i], false))
+                    .unwrap_or_else(|| (initial_key, false)),
+            };
+
+            if (bad_substitution
+                || key
+                    .chars()
+                    .any(|c: char| !c.is_ascii_alphanumeric() && c != '_'))
+                && !is_escaped(prefix)
+            {
+                return Some(Warning::new(
+                    line.clone(),
+                    self.name,
+                    self.message(line.get_key()?),
+                ));
+            }
+
+            value = &rest;
+        }
+        None
+    }
+
+    fn name(&self) -> &str {
+        self.name
+    }
+}
+
+impl SubstitutionKeyChecker<'_> {
+    fn message(&self, key: &str) -> String {
+        self.template.replace("{}", &key)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::common::tests::*;
+
+    fn run_substitution_tests(asserts: Vec<(LineEntry, Option<Warning>)>) {
+        let mut checker = SubstitutionKeyChecker::default();
+
+        for assert in asserts {
+            let (input, output) = assert;
+            assert_eq!(checker.run(&input), output);
+        }
+    }
+
+    #[test]
+    fn correct_substitution_key_test() {
+        let asserts = vec![
+            (line_entry(1, 3, "ABC=$BAR"), None),
+            (line_entry(2, 3, "FOO=${BAR}"), None),
+            (line_entry(3, 3, "FOO=\"$BAR\""), None),
+        ];
+
+        run_substitution_tests(asserts);
+    }
+
+    #[test]
+    fn incorrect_substitution_key_test() {
+        let asserts = vec![
+            (
+                line_entry(1, 2, "ABC=${BAR"),
+                Some(Warning::new(
+                    line_entry(1, 2, "ABC=${BAR"),
+                    "SubstitutionKey",
+                    "The ABC key is not assigned properly",
+                )),
+            ),
+            (
```

Comment:
Also, we should cosider `ABC=$BAR}` as invalid substitution key (because they potentially forgot to specify `{`).
Please, add this case 🙂 

```suggestion
            ),
            (
                line_entry(2, 2, "FOO=$BAR}"),
                Some(Warning::new(
                    line_entry(2, 2, "FOO=$BAR}"),
                    "SubstitutionKey",
                    "The FOO key is not assigned properly",
                )),
            ),
            (
```

Reply:
That is supposed to work I believe

Reply:
> That is supposed to work I believe

With the approach I proposed below, it definitely works. But not in the current one, if I'm not mistaken

Reply:
I meant to say that in bash, it is such:
```bash
$ BAR="hello"
$ FOO=$BAR}
$ echo $FOO
hello}
```

Reply:
In this case, the linter's task is to identify even potential warnings.  So as described in the ischew, I would rather consider this as `Warning` (as described in issue) 🤔